static/pane.ts

```diff
@@ -0,0 +1,195 @@
+        this.domRoot = container.getElement();
+        this.initializeDOMRoot();
+
+        const editorRoot = this.domRoot.find('.monaco-placeholder')[0];
+        this.createEditor(editorRoot);
+
+        this.selection = state.selection
+        this.compilerInfo = {
+            compilerId: state.id,
+            compilerName: state.compilerName,
+            editorId: state.editorid,
+        }
+        this.fontScale = new FontScale(this.domRoot, state, this.editor);
+        this.topBar = this.domRoot.find('.top-bar');
+
+        this.registerButtons(state);
+        this.registerStandardCallbacks();
+        this.registerCallbacks();
+        this.setTitle();
+    }
+
+    /**
+     * Initialize the DOM root node with the pane's default content. Typical
+     * implementation looks like this:
+     *
+     * ```ts
+     * this.domRoot.html($('#rustmir').html());
+     * ```
+     */
+    abstract initializeDOMRoot(): void;
+
+    /**
+     * Initialize the monaco editor instance. Typical implementation for looks
+     * like this:
+     *
+     * ```ts
+     * this.editor = monaco.editor.create(editorRoot, extendConfig({
+     *     // goodies
+     * }));
+     * ```
+     */
+    abstract createEditor(editorRoot: HTMLElement): void;
+
+    /**
+     * Emit analytics event for opening the pane tab. Typical implementation
+     * looks like this:
+     *
+     * ```ts
+     * ga.proxy('send', {
+     *   hitType: 'event',
+     *   eventCategory: 'OpenViewPane',
+     *   eventAction: 'RustMir',
+     * });
+     * ```
+     */
+    abstract registerOpeningAnalyticsEvent(): void
+
+    /** Optionally overridable code for initializing pane buttons */
+    registerButtons(state: BasePaneState /* typeof state */): void {}
+
+    /** Optionally overridable code for initializing event callbacks */
+    registerCallbacks(): void {}
+
+    abstract getPaneName(): string;
+    abstract onCompiler(id: number, compiler: any, options: any, editorId: number): void;
+    abstract onCompileResult(id: unknown, compiler: unknown, result: unknown): void;
+    abstract close(): void;
+
+    /** Initialize standard lifecycle hooks */
+    protected registerStandardCallbacks(): void {
+        this.fontScale.on('change', () => this.updateState());
+        this.container.on('destroy', () => this.close());
+        this.container.on('resize', () => this.resize());
+        this.eventHub.on('compileResult', (id, compiler, result) => {
+            this.onCompileResult(id, compiler, result);
+        });
+        this.eventHub.on('compiler', (id, compiler, options, editorId) => {
+            this.onCompiler(id, compiler, options, editorId);
+        });
+        this.eventHub.on('compilerClose', (id) => this.onCompilerClose(id));
+        this.eventHub.on('settingsChange', (settings) => this.onSettingsChange(settings));
+        this.eventHub.on('shown', () => this.resize());
+        this.eventHub.on('resize', () => this.resize());
+    }
```

Comment:
These events are pretty much standard for each pane, so registering them in a different place reduces code needed for implementors.